### ISSUE TITLE
Discussion of a wild idea: Multiple Genesis Blocks, Archival Node

### ISSUE THREAD

**Issue by Technologov** [dashpay/dash#1380]
I think we may be able to reduce a block chain size dramatically, by re-generating automatically a new genesis block every year, (or every X blocks), and keeping the balance of the previous block chain, last block of it.
Theoretically it will solve the hard disk space issues, And will reduce network bandwidth requirements for setting a new Full node. It will not solve new block propagation delays.

2016 [block 0]...[block n] --- the last block's balance could transfer into a new genesis block 2017.
2017 [block 0]

Afterwards Mining should start build up on the new chain. Just the history of transactions will be lost.

There could be more types of nodes: Archival node, keeping all history of all the previous block chains, and the Full Node, keeping only the current block chain.

In this scenario, only block explorers and scientists will need to keep an Archival Node.

Do you think it is doable, theoretically?

This idea, if it works, will mean that Dash will beat Visa x 10 transactions easily, with a goal to break 1 billion transactions per day, in a decentralized way.

04.Mar.2017

-Technologov

**Comment by Technologov** [comment]
Clarification 01: New genesis block must be generated in a deterministic way, to be reproducible by all miners and Masternodes. 

Clarification 02: tiny dust amounts  can be distributed to the last 1000 miners. 
Let's define 'dust' as transaction outputs  below the median fees for the last 100 blocks.

**Comment by akhavr** [comment]
1. How fresh new client would determine if he's fed with a valid masterblock
2. If the validity of a masterblock would be confirmed by masternodes, how fresh new client would determine which masternodes to trust?

**Comment by Technologov** [comment]
It's already answered in C04 (Clarification 04) and C05.

**Comment by taw00** [comment]
I think there are some ideas swirling around the bitcoin world on this topic. Part of the challenge is that certain types of transactions have to be honored somehow: Future dated, etc.

**Comment by sidhujag** [comment]
You may want to look into MerkleDAGs because they essentially use the last transaction which is a hash of all the previous hash's as the validity of any future transactions, essentially a blockchain-less system that tries to emulate a chain of transactions to do the same thing.

**Comment by Technologov** [comment]
Is there any coins that implement said "Directed acyclical Merkle DAGs" for tx scalability?

**Comment by sidhujag** [comment]
Iota and to lesser extent byteball afaik.. its still unproven however read their paper and critique it

**Comment by Technologov** [comment]
Taw00: maybe we can make a special section in the MB with special transactions, and list them as 'valid' - Masternode tx with 1000.0 DASH output, future dated, multisig, etc... it needs some though.

**Comment by Technologov** [comment]
Opinion from Bitcoin people:
https://bitcointalk.org/index.php?topic=1851463.0

spartacusrex said: (April 03, 2017)

> Like. 
> 
> BUT. Could be easier. Don't need a Master Block.
> 
> 1) Throw away EVERYTHING except the Block-header for blocks older than 1 day / (..or the longest re-org imaginable ). The block header is the ONLY bit you really need - as it has the POW. Also it links back to the previous block header, so we have our 'Proof Chain' back to the Genesis block. (Which would be tiny). But still packed with ridiculous levels of POW-er..
> 
> 2) You need to store the deterministic UTXO root hash somewhere in the block header. Changes for every block obviously. But everyone would agree on  it.
> 
> 3) That's it.
> 
> As with your design we lose the TXN history - but we keep the POW chain that led us to the present day UTXO.
> 
> So.. doing the Maths.. it would be about :
> 
> 576GB (as you calculated for 1 billion txn a day)+ UTXO database size (orders of magnitude smaller than 0.576TB) + the proof chain size (tiny).. forever.
> 
> Whether there would be any point to keeping the TXN history - is debatable..

**Comment by andyfreer** [comment]
this seems like a really inefficient way to prune down to UTXO set.  The masterblock would be huge - nodes can just prune if they want.  There are no real savings here, and pruning spent tx on all nodes will cause other issues an lose history e.g. metadata in spent tx.

**Comment by Technologov** [comment]
Standard blockchain pruning, as done in Bitcoin disallows people to start new nodes from pruned nodes.

**Comment by andyfreer** [comment]
yes but new nodes synching from a masterblock are gaining the same info as they would from a pruned node. in fact less.  

the main issues are i have are:

1. this is just an ugly way to use blocks - the network has a certain ebb and flow with blocks gradually adding new tx to the chain every 2.5 minutes - shoving everything in a gigantic block once a year is not within the design rationale for such a system - apart from needing an absolutely huge block (an entire year of UTXO in one block) that needs to be propagated and verified, your adding this huge focal point where if that block is attacked the network is stuck

2. to do this we have to be ok with pruning everything except UTXO. the reason this is optional is there are TX types and metadata that are really needed.  Moving that problem to archive nodes just reduces the security, and how will they be incentivized etc?

Sorry to be a wet wrag, just think this needs a lot more thought and deeper understanding of the architecture.  Maybe one of the core devs can answer better.

Andy

**Comment by Technologov** [comment]
1. It will increase delay of the Dash network once a year. Ugly ? Yes, a little bit. But it is better than having a multi-petabyte storage array, that even incentivized Masternodes will be unable to afford.

2. I think special transactions (such as time-locked utxo, masternode tx with 1000.0 DASH output and multisig) should be put into a special "extended space" inside a Masterblock, and special transactions must be marked as valid as long as an address has the required balance.

=========
After all: We have several years to figure out a long-term scalability strategy. But it needs to be solved before Dash, or any crypto-currency for that matter, could become mass adopted by a billion users.
My proposal is one possible solution, maybe even not the best, and I urge everyone to look for solutions, because this problem is *critical*. But Masterblocks idea is a solution worth keeping in mind, if all else fails.

**Comment by andyfreer** [comment]
1. Ok but right now the highest load we know works is 1MB per block every 10 minutes (excluding Dash testnet which test ok at 4x that and pretty sure we can push that far higher).  How big would the masterblock be?  Could be 1GB for the year on 10GB chain growth?  How is that possibly going to be propagated / verified in a timely / secure fashion, assuming we remove any kind of size limit?  It's also a huge attack service at a given time that can be targetted.  The MB size also scales with the average block size so it would get larger as blocks get larger

2. But then we need to discern which spent tx to add to the masterblock making it larger

There are more elegant solutions being researched in the space like sharding (although sceptical on that too).  Basically, don't see any elegant solutions to this yet.  Lot of solutions in the space try to improve on Bitcoin efficiency but truth is the core architecture is about as efficient you can get without some radical redesign. 

To make this viable, i think you'd need to first show that a masterblock is feasible in itself - size, propagation, verification time, security with such a larger block at predesignated height.  but i'm pretty sure you will find that it's not possible within the architectural limits IIRC.

**Comment by joezippy** [comment]
Hello,

I know I'm new here, but I might have something you want to consider, as you move forward with the concept of "more types of nodes: Archival node, keeping all history of all the previous block chains, and the Full Node, keeping only the current block chain." (@Technologov)... 

The idea (more details in the forum post) has to do with speeding up the mixing of coins for Private Send functionality, essentially by flagging transactions on the chain, as already "mixed / private" previously...  Maybe this is a "PrivateSend node" or just identified on the "Archival node" as type private... Regardless... The idea also suggests moving the private send / mixing functionality from the wallet endpoints, to the masternodes which would allow all wallets to PrivateSend, not just the QT in the future.... 

Anyway, here is the thread on the dash.org forum... It should probably be moved out of the "General Discussion" to the "Development Tech Discussion" for more developer visibility, but I can't do that... :)

https://www.dash.org/forum/threads/masternode-private-send-mixing.14238/

Let me know if you have questions about the ideas... I'm not at the same level as you all are (yet), but I'm on the same Dash team!

Thanks for your time... Cheers!

**Comment by joshafest** [comment]
The idea is good!! 
Can anyone add features and create pull here ?
I mean can anyone submit his work by creating pull request or it is only done by team ?

**Comment by Technologov** [comment]
Anyone can submit a patch, that will be reviewed by the Dash Core team.

**Comment by joezippy** [comment]
@joshafest ... Just for clarification is it my idea you think, is good for resolving current issues w/ mixing and private send or the bigger topic @Technologov is proposing... If it's mine, are you considering implementing this pseudo-code?

```
// So I might be way off base, but somewhere in the dApp lives a 'write' function 
// and somewhere their lives a 'read'... If that is true... Maybe something like this 
// could be done...

public read(addr) {
   // psAddressRegistry contains all address that are private 
   if (addr in psAddressReqistry) {
       addr.trans.obfuscate.sender()
       addr.trans.obfuscate.receiver()
   }
}

public write(addr) {
   addr.trans.do-what-you-do-today()
}
```

Wishing I had the skills to implement it myself... Just not there anymore... I'm hopeful you can get it done @joshafest !  :) 

Cheers!

**Comment by Technologov** [comment]
joezippy: Private send is NOT part of my idea. This idea -- "Masterblocks" -- is squarely targeted at scalability problems, not privacy.

**Comment by joezippy** [comment]
@Technologov not trying to hijack the thread, just clarify @joshafest comment... Cheers!

**Comment by joshafest** [comment]
What I think is .. accept the last checkpoint block as initial block with height and treat it as it was genesis block ...
Problems
- one who have coins on much earlier blocks, have no effect of this solution as he has to download whole chain.
-One who want to import private key, have problems.

**Comment by arminnaderi** [comment]
@Technologov read [https://github.com/cyberdust/core-node/wiki/White-Paper](url)

**Comment by Technologov** [comment]
I have updated the specs:
https://docs.google.com/document/d/1usnNigTJw6cpone_xrcIMmzxF7U1MFrjb3y7aTsbDx0/edit?usp=sharing
